cgi-bin/DW/Console/Command/BonusIcons.pm

```diff
@@ -56,25 +58,54 @@ sub execute {
     }
 
     return $self->error( 'Invalid subcommand.' )
-        if $cmd && $cmd !~ /^(?:add|remove)$/;
+        if $cmd && $cmd !~ /^(?:add|remove|xfer)$/;
 
-    return $self->error( 'Count must be a positive integer.' )
-        unless $count =~ /^\d+$/;
-    $count += 0;
+    if ( $cmd eq 'add' || $cmd eq 'remove' ) {
 
-    if ( $cmd eq 'add' ) {
-        my $new = max( $to_u->prop( 'bonus_icons' ) + $count, 0 );
-        $to_u->set_prop( bonus_icons => $new );
-        LJ::statushistory_add( $to_u, $remote, 'bonus_icons',
-                sprintf( 'Added %d icons, new total: %d.', $count, $new ) );
-        $self->print( sprintf( 'User now has %d icons.', $new ) );
+        my $count = $cmdarg;
+
+        return $self->error( 'Count must be a positive integer.' )
+            unless $count =~ /^\d+$/ || $cmd eq 'xfer';
+        $count += 0;
+
+        if ( $cmd eq 'add' ) {
+            my $new = max( $to_u->prop( 'bonus_icons' ) + $count, 0 );
+            $to_u->set_prop( bonus_icons => $new );
+            LJ::statushistory_add( $to_u, $remote, 'bonus_icons',
+                    sprintf( 'Added %d icons, new total: %d.', $count, $new ) );
+            $self->print( sprintf( 'User now has %d icons.', $new ) );
+
+        } elsif ( $cmd eq 'remove' ) {
+            my $new = max( $to_u->prop( 'bonus_icons' ) - $count, 0 );
+            $to_u->set_prop( bonus_icons => $new );
+            LJ::statushistory_add( $to_u, $remote, 'bonus_icons',
+                    sprintf( 'Removed %d icons, new total: %d.', $count, $new ) );
+            $self->print( sprintf( 'User now has %d icons.', $new ) );
+
+        }
+
+    } elsif ( $cmd eq 'xfer' ) {
+        my $target_u = LJ::load_user( $cmdarg );
+
+        return $self->error( 'Invalid target user.' )
+            unless $target_u;
+        return $self->error( 'E-mail addresses do not match.' )
+            unless $to_u->prop( 'email' ) eq $target_u->prop( 'email' ); 
 
-    } elsif ( $cmd eq 'remove' ) {
-        my $new = max( $to_u->prop( 'bonus_icons' ) - $count, 0 );
-        $to_u->set_prop( bonus_icons => $new );
+        my $xfer_count = $to_u->prop( 'bonus_icons' );
+        my $new_to_u = max( $to_u->prop( 'bonus_icons' ) - $xfer_count, 0 );
```

Comment:
So in fact this is also the point where I have a question about how to simplify the maths -- I think I had a chat about how Perl handles objects and things that point at them with geekosaur in IRC when I was working on this, but what I'm tripping up on here is:

$xfer_count is the full number of bonus icons that $to_u has. $to_u->prop( 'bonus_icons' ) needs setting to 0 (which... i could actually just do rather than faffing with the subtraction, okay, well done me) without making $xfer_count also 0. If I set $xfer_count, then reset $to_u->prop( 'bonus_icons' ) to 0, in Perl does $xfer_count stay the number I first thought of (i.e. the number I want to add to $target_u)? (I tried reading docs and got myself confused on this point. :-/ Happy to be pointed at what I should be reading, though!)